doc/source/architecture.rst

```diff
@@ -0,0 +1,213 @@
+Invocation.
+
+Message passing
+---------------
+
+Message passing is the original model of remote APIs, and in fact the other two
+are implemented using message passing underneath. RPCs are initiated by the
+client and sent to the server as a message, or a sequence of bytes. There must
+be some understanding on both sides of the channel as to what information is in
+these bytes. This is usually handled by a protocol (such as HTTP or TCP) and
+usually also a higher level library (such as gRPC or zeroMQ) implemented using
+the protocol. The server may send back a series of messages to the client to
+indicate that a message was received, that it is executing, that it has
+executed, and with a response, if any. The protocol and library determine the
+expected sequence of messages, their ordering, and any failure handling.
+
+Resource-based APIs
+-------------------
+
+Resource-based APIs were popularized by REST, and has been shown to be, for
+practical purposes, infinitely scalable. They are traditionally implemented
+using HTTP or HTTPS but implementations can exist in any transport protocol.
+
+The fundamental concept behind REST is the separation of verbs and resources.
+Verbs include GET, PUT, UPDATE, DELETE, and POST, while resources are any
+uniquely identifiable entity.
+
+REST can scale because servers can make assumptions about the data it serves
+based on the verb and resource. For instance, if a GET was done on the resource
+"/a/b/c", and then no mutating verbs (PUT, UPDATE, DELETE, POST) have been run
+on that or any child resource (such as "/a/b/c/d"), the server can reuse the
+result of the previous request rather than recompute the result. When using the
+HTTP protocol, this is called HTTP caching and is a fundamental property of the
+internet.
+
+Remote method invocation
+------------------------
+
+In the 1990s, OOP exploded in popularity. With it, came a very tempting idea.
+OOP allows programmers to provide abstractions on top of data in their code.
+It was thought that even the difference between RPC calls and calls made in
+a program's address space could be abstracted. The user of the library does
+not need to know whether an object exists remotely or locally, the call can
+just be made in the same way, and to scale the system to be distributed over a
+network, some middleware would be responsible for load-balancing and allocating
+these objects remotely.
+
+This technique is known as Remote Method Invocation (RMI). RMI was widely
+implemented, using CORBA, DCOM, Remoting (.NET), and Java RMI. However, this
+approach has fallen out of favor with the rise of the internet, as it was
+observed that this paradigm does not scale. As a practical example, recent
+versions of .NET do not implement the Remoting library, and COM/DCOM are not
+used in modern web stacks.
+
+An illuminating discussion of the problems with RMI can be found at
+https://martinfowler.com/articles/distributed-objects-microservices.html
+
```

Comment:
The advent of OOP enabled programmers to provide abstraction on top of data in their code. They could hide the message-passing of the network, that is, the messages between RPC calls and calls made in a program's address space. Using this configuration, the user of the library does not need to know whether an object exists remotely or locally. This includes the call. Additionally, you can distribute the system over a network and use a middleware application to manage the necessary load-balancing and allocation of these objects remotely. Known as Remote Method Invocation (RMI), developers implemented this technique using CORBA, DCOM, Remoting (.NET), and Java RMI.
Unfortunately, based on scalability limitations, the favorability of RMI has waned. Recent versions of .NET do not implement the Remoting library, and modern web stacks  do not use COM/DCOM.
This article discusses the problems associated with RMI: https://martinfowler.com/articles/distributed-objects-microservices.html.